postgis/mvt.c

```diff
@@ -687,108 +687,99 @@ LWGEOM *mvt_geom_fast(LWGEOM *lwgeom, const GBOX *gbox, uint32_t extent, uint32_
 	lwgeom_simplify_in_place(lwgeom, res, preserve_collapsed);
+
 	/* If geometry has disappeared, you're done */
 	if (lwgeom_is_empty(lwgeom))
 		return NULL;
 
 	if (clip_geom)
 	{
-		const GBOX *ggbox = lwgeom_get_bbox(lwgeom);
-		GBOX bgbox = *gbox;
+		GBOX bgbox;
+		const GBOX *lwgeom_gbox = lwgeom_get_bbox(lwgeom);;
+		bgbox = *gbox;
 		gbox_expand(&bgbox, buffer_map_xunits);
-		if (!gbox_overlaps_2d(ggbox, &bgbox))
+		if (!gbox_overlaps_2d(lwgeom_gbox, &bgbox))
 		{
-			POSTGIS_DEBUGF(3, "%s: geometry outside clip box", __func__);
+			POSTGIS_DEBUG(3, "mvt_geom: geometry outside clip box");
 			return NULL;
 		}
-		if (!gbox_contains_2d(&bgbox, ggbox))
+		if (!gbox_contains_2d(&bgbox, lwgeom_gbox))
 		{
 			double x0 = bgbox.xmin;
 			double y0 = bgbox.ymin;
 			double x1 = bgbox.xmax;
 			double y1 = bgbox.ymax;
-#if POSTGIS_GEOS_VERSION < 35
-			LWPOLY *lwenv = lwpoly_construct_envelope(0, x0, y0, x1, y1);
-			lwgeom_out = lwgeom_intersection(lwgeom, lwpoly_as_lwgeom(lwenv));
-			lwpoly_free(lwenv);
-#else
-			lwgeom_out = lwgeom_clip_by_rect(lwgeom, x0, y0, x1, y1);
-#endif
-			POSTGIS_DEBUGF(3, "%s: no geometry after clip", __func__);
-			if (lwgeom_out == NULL || lwgeom_is_empty(lwgeom_out))
+			lwgeom = lwgeom_clip_by_rect(lwgeom, x0, y0, x1, y1);
+			POSTGIS_DEBUG(3, "mvt_geom: no geometry after clip");
+			if (lwgeom == NULL || lwgeom_is_empty(lwgeom))
 				return NULL;
 		}
 	}
 
-	/* If no clipped output we will use raw input */
-	/* This means we require the input to build on top */
-	/* of a PG_DETOAST_DATUM_COPY() */
-	if (!lwgeom_out)
-		lwgeom_out = lwgeom;
-
-	/* if polygon(s) force clockwise as per MVT spec */
-	if (lwgeom_out->type == POLYGONTYPE ||
-		lwgeom_out->type == MULTIPOLYGONTYPE) {
-        lwgeom_force_clockwise(lwgeom_out);
-	}
-
 	/* transform to tile coordinate space */
 	memset(&affine, 0, sizeof(affine));
 	affine.afac = fx;
 	affine.efac = fy;
 	affine.ifac = 1;
 	affine.xoff = -gbox->xmin * fx;
 	affine.yoff = -gbox->ymax * fy;
-	lwgeom_affine(lwgeom_out, &affine);
+	lwgeom_affine(lwgeom, &affine);
 
 	/* snap to integer precision, removing duplicate points */
 	memset(&grid, 0, sizeof(gridspec));
 	grid.ipx = 0;
 	grid.ipy = 0;
 	grid.xsize = 1;
 	grid.ysize = 1;
-	lwgeom_grid_in_place(lwgeom_out, &grid);
+	lwgeom_grid_in_place(lwgeom, &grid);
 
-	if (lwgeom_out == NULL || lwgeom_is_empty(lwgeom_out))
+	if (lwgeom == NULL || lwgeom_is_empty(lwgeom))
 		return NULL;
 
-	/* if polygon(s) force valid as per MVT spec */
-    // if (lwgeom_out->type == POLYGONTYPE ||
-    //     lwgeom_out->type == MULTIPOLYGONTYPE) {
-    //     /* Watch out, does make_valid try to force orientation? */
-    //     lwgeom_out = lwgeom_make_valid(lwgeom_out);
-    // }
+	/* if polygon(s) make valid and force clockwise as per MVT spec */
+	if (lwgeom->type == POLYGONTYPE ||
+		lwgeom->type == MULTIPOLYGONTYPE ||
+		lwgeom->type == COLLECTIONTYPE)
+	{
+		lwgeom = lwgeom_make_valid(lwgeom);
```

Comment:
Sure you want to bring this back? Validity forcing is the major cause of ST_AsMVT slowness in my early profiles of it, and the primary reason the _fast variant exists.

Reply:
In this iteration, yes. The idea is first to have it passing all tests, then start optimizing maintaining valid output. Then all this should land both here and in trunk.

Reply:
Checking the spec I see that invalid polygons are not supported:
```
Linear rings MUST be geometric objects that have no anomalous geometric points, such as self-intersection or self-tangency. The position of the cursor before calling the ClosePath command of a linear ring SHALL NOT repeat the same position as the first point in the linear ring as this would create a zero-length line segment. A linear ring SHOULD NOT have an area calculated by the surveyor's formula equal to zero, as this would signify a ring with anomalous geometric points.
```

What's more, I've tested one of them using Mapnik (`(SELECT 2 AS cartodb_id, GeomFromEWKT('SRID=4326;POLYGON((0 85, 150 0, -150 0, 0 -85, 0 85))') as the_geom, FALSE as status2, 0 as data) AS cdbq2`) and it's simply parsed as empty (completely ignored) so it's certainly something where we could chose to either adapt them (`lwgeom_make_valid`) or drop them completely, but I don't think we can leave them as they are since that isn't allowed in the spec.

Reply:
Right, invalids are not to spec, but fixing (or detecting) them is immensely expensive. Also, they cannot be pre-flagged and pre-fixed since the very act of simplifying and gridding the geometry may render them invalid during the processing. It makes generating "on the fly" MVT very hard to do while remaining to spec and not burning piles of time checking every polygon for invalidity.

Reply:
Or we could just add an alert in the docs to tell people that the input should be valid or otherwise the mvt might (will?) be invalid?

Reply:
Valid input does not guarantee valid output, as we are moving the vertices (snap) and edges (simplify) around in ways that can generate invalid output from valid input.

Reply:
I think we can get away with invalid output with invalid input, but we shouldn't generate invalid output with valid input (Mapnik doesn't as far as I know).

>  we are moving the vertices (snap)
We could just force valid if snapping (which we don't do). Also, since both are geos operations we could have a function that does the snapping and forces validity to avoid unnecessary transformations .

> and edges (simplify) 
Geos at least is able to simplify returning valid geometries (DouglasPeuckerSimplifier) so we could use that.

Maybe it'd be more efficient if everything is done using geos directly (not sure if possible). Do you think it's worth a try or would you rather go with the invalid outputs for certain cases?

Reply:
Sorry, moving vertices when gridding, which we do to everything.

My assumption in stripping out the validity checks was that the penalty for invalidity would be an occasional mis-rendering, and compared to the known performance penalty for *checking or fixing* invalidity on every geometry, an occasional mis-rendering was a small price to pay.